lib/rucio/daemons/conveyor/preparer.py

```diff
@@ -0,0 +1,224 @@
+    heartbeat.sanity_check(executable=executable, hostname=hostname, pid=pid, thread=current_thread)
+
+    try:
+        graceful_stop.wait(10)  # gathering of daemons/threads on first start
+        while not graceful_stop.is_set():
+            start_time = time()
+
+            pulse = heartbeat.live(executable=executable, hostname=hostname, pid=pid, thread=current_thread)
+            worker_number = pulse['assign_thread']
+            total_workers = pulse['nr_threads']
+
+            try:
+                updated_msg = run_once(total_workers=total_workers, worker_number=worker_number, limit=bulk)
+            except RucioException:
+                logging.exception('conveyor-preparer[%s/%s] errored with a RucioException, retrying later' % (worker_number, total_workers))
+                updated_msg = 'errored'
+
+            if once:
+                break
+
+            end_time = time()
+            time_diff = end_time - start_time
+            logging.info('conveyor-preparer[%s/%s] %s, taking %.3f seconds' % (worker_number, total_workers, updated_msg, time_diff))
+            if time_diff < sleep_time:
+                sleep_remaining = sleep_time - time_diff
+                logging.info('conveyor-preparer[%s/%s] sleeping for a while :  %.2f seconds' % (worker_number, total_workers, sleep_remaining))
+                graceful_stop.wait(sleep_remaining)
+
+        logging.info('conveyor-preparer[%s/%s]: gracefully stopping' % (worker_number, total_workers))
+
+    finally:
+        heartbeat.die(executable=executable, hostname=hostname, pid=pid, thread=current_thread)
+
+
+def run_once(total_workers=0, worker_number=0, limit=None, session: "Session" = None) -> str:
+    req_sources = __list_transfer_requests_and_source_replicas(
+        total_workers=total_workers,
+        worker_number=worker_number,
+        limit=limit,
+        request_state=RequestState.PREPARING,
+        session=session
+    )
+    if not req_sources:
+        return 'had nothing to do'
+
+    # sort by Request.id, should be pretty quick since the database should return it this way (tim sort)
+    req_sources = sorted(req_sources, key=lambda t: t[request_id_col])
+
+    def minimum_distance_request():
+        # req_sources must be checked at this point, see above
+        cur_request_id = req_sources[0][request_id_col]
+        shortest_item = req_sources[0]
+        for idx in range(1, len(req_sources)):
+            if cur_request_id != req_sources[idx][request_id_col]:
+                yield shortest_item
+                cur_request_id = req_sources[idx][request_id_col]
+                shortest_item = req_sources[idx]
+            elif req_sources[idx][distance_col] < shortest_item[distance_col]:
+                shortest_item = req_sources[idx]
+        yield shortest_item
+
+    count = _update_request_chunk(minimum_distance_request(), session=session)
+    return f'updated {count}/{limit} requests'
+
+
+@transactional_session
```

Comment:
Shouldn't it be moved to core ?

Reply:
Good question. It is preparer-specific though - I feel like it would be misplaced in core.

Reply:
It's debatable, but in general we tried to put everything with database interactions to the core, even if only used by the one daemon. (It's the same for the judges etc.)
I would move it.

Reply:
will do